Automattic/wp-calypso
# Stepper: Add unified domain step

Context: pdDR7T-1Ed-p2#comment-2179

## Proposed Changes

This aims to create the thinnest wrapper possible to wrap the Domains step in /start and have it in Stepper as well, to achieve parity without repeating any work. 

## Why are these changes being made?
To be able to delete `client/landing/stepper/declarative-flow/internals/steps-repository/domains` and `client/landing/stepper/declarative-flow/internals/steps-repository/choose-a-domain` and have a single Domains step across all frameworks and places. 

It's still having an issue of handling the picked domain. Needs more work. 

## Testing Instructions

1. Go to /setup/onboarding.
2. Play with the domain step.
3. It should be identical to the one at /start/domains.
4. It should allow selecting multiple or one domain domains.
5. The flow shouldn't take you to checkout if you pick a free domain.
6. The flow SHOULD take you to checkout if you pick one or more paid domains.
7. It should look identical on mobile too.

### Back button
1. If you go to the /setup/onboarding flow without owning any sites, there shouldn't be back button in the domains step.
2. If you go to  /setup/onboarding while owning one site, the back button should say "Back to My Home".
3. If you go to /setup/onboarding while owning two or more sites, back button should say "Back to sites" and takes you to /sites.

### Regression testing
1. Go to https://wpcalypso.wordpress.com/setup/site-setup/goals?siteSlug=YOUR_SITE.
2. Make sure the back button works great.

### What's not to test

1. The interaction with the plans grid. The plans grid isn't yet compatible with this domain step, so it doesn't know if you picked a paid domain or such. Don't test the plans grid, just the domains step, please.

## Next steps

1. ~~Find a way to store the selected domain in the Onboarding store in away that accessible to the plans grid. Similar to `client/landing/stepper/declarative-flow/internals/steps-repository/domains`~~ Done.
5. ~~Find a typed way to persist the [step state](https://github.com/Automattic/wp-calypso/blob/add/unified-domain-step/client/landing/stepper/declarative-flow/internals/steps-repository/unified-domains/index.tsx#L96).~~ Done. 
9. Test test test.
10. ~~Consider doing the same thing for the plans step.~~ Done.
